src/Compilers/CSharp/Test/Semantic/Diagnostics/GetDiagnosticsTests.cs

```diff
@@ -555,5 +556,60 @@ static async Task verifyDiagnosticsAsync(Compilation compilation, DiagnosticAnal
                 Assert.False(diagnostic.IsSuppressed);
             }
         }
+
+        [Fact]
+        public async Task TestConcurrentGetAnalyzerDiagnostics()
+        {
+            var source1 = @"
+partial class C
+{
+    void M1()
+    {
+        // warning CS0168:  The variable 'x' is declared but never used.
+        int x;
+    }
+}
+";
+            var source2 = @"
+partial class C
+{
+    void M2()
+    {
+        // warning CS0168:  The variable 'x' is declared but never used.
+        int x;
+    }
+}
+";
+            var source3 = @"
+class C3
+{
+    void M2()
+    {
+        // warning CS0168:  The variable 'x' is declared but never used.
+        int x;
+    }
+}
+";
+            var compilation = CreateCompilation(new[] { source1, source2, source3 });
+            compilation = compilation.WithOptions(compilation.Options.WithConcurrentBuild(true));
+
+            var analyzers = ImmutableArray.Create<DiagnosticAnalyzer>(new CSharpCompilerDiagnosticAnalyzer());
+            var compilationWithAnalyzers = compilation.WithAnalyzers(analyzers,
+                new CompilationWithAnalyzersOptions(
+                    new AnalyzerOptions(ImmutableArray<AdditionalText>.Empty),
+                    onAnalyzerException: null,
+                    concurrentAnalysis: true,
+                    logAnalyzerExecutionTime: false));
+
+            var tree = compilation.SyntaxTrees.First();
+            var model = compilation.GetSemanticModel(tree, true);
+            var tasks = new Task[10];
+            for (var i = 0; i < 10; i++)
+            {
+                tasks[i] = Task.Run(() => compilationWithAnalyzers.GetAnalyzerSemanticDiagnosticsAsync(model, null, CancellationToken.None));
+            }
+
+            await Task.WhenAll(tasks);
```

Comment:
>await Task.WhenAll(tasks); [](start = 12, length = 26)

Do we need to verify anything about the analysis? How do we know any diagnostics were generated?

Reply:
How did this test fail previously?

---
In reply to: [462691456](https://github.com/dotnet/roslyn/pull/45607#discussion_r462691456) [](ancestors = 462691456)

Reply:
Yes, AnalyzerManager.VerifySymbolEndActionsExecuted asserts as instance of analyzer manager has corrupted state. Also attempting to request semantic diagnostics for a file from different threads causes prior pending requests to be singnalled for cancellation and leads to unhandled Operation canceled exception